pkg/name/digest.go

```diff
@@ -79,18 +88,23 @@ func NewDigest(name string, opts ...Option) (Digest, error) {
 		return Digest{}, err
 	}
 
+	var writeTarget Reference = nil
 	tag, err := NewTag(base, opts...)
 	if err == nil {
 		base = tag.Repository.Name()
+		if tag.explicitTag() {
+			writeTarget = tag
```

Comment:
Since we discard the digest from the write target, write operations need to validate that the digest matches that of the image and return an error if it does not.